### DISCUSSION TITLE
Discussion: Search and Icons

### DISCUSSION THREAD

**Discussion by sturobson** [visual-framework/vf-core#1388] [Proposals]
It was briefly discussed in our recent VF/UX Chat (23/02/21) about the possible need for an icon for the search input in certain instances.

Here's the stub for a discussion on the topic.

As we are moving to using the streamline icon set for `2.x.x`+ I've dropped in the relative icons that 'might fit'.

Note: We are tending to use the 'bold' versions of icons, so I've only added those options here:

![search-1](https://user-images.githubusercontent.com/925197/108979639-c5208180-7682-11eb-9efb-2175c4061f4e.png)

![search-alternate](https://user-images.githubusercontent.com/925197/108979641-c5b91800-7682-11eb-950f-cf97983c91a3.png)

There are options for search icons in boxes and circles - but this would go against the style used in existing icons.

But I can share those if needed/requested.

**Comment by sturobson** [comment]
A quick note is that with the new Visual Framework + we are trying not rely on icons where labels will work, and work for a wider audience.

**Comment by khawkins98** [comment]
Just adding a note that echoes @sturobson's: adding icons will certainly provide some benefit in certain cases, however it also adds questions around governance, UX, visual and technical questions will distract from other work.

My current take is that we should consider exploring the role that icons could/should play in the VF, but that for implementation we [focus on the existing roadmapped milestones](https://github.com/visual-framework/vf-core/milestones)

**Comment by gabsie** [comment]
Hi, both @khawkins98 and @sturobson 

I suggested to use this in Search fields which act as filters, as opposed to Search boxes which are acting as Search, where they are usually accompanied by the Search button. 
I and others have seen that pattern elsewhere, and provides good feedback.

@sturobson - maybe we can see those in the field itself and decide then if people are OK with this functionality

Gabs

**Reply by sturobson** [reply]
Hey @gabsie - so the icon is to show that it's an 'interactive' type ahead style search filter. 

I've noticed that the govuk design system do this:

![Screenshot 2021-02-24 at 19 11 41](https://user-images.githubusercontent.com/925197/109053671-e0b47800-76d4-11eb-9f9a-f1ab5f30fc0b.png)

Myself, @cindyebi and @nikiforosk were discussing form fields again today and we happened upon the idea that 'search is a special case' for 'placeholder text' (the text that's inside the input until the user starts typing. Would that be an option?

We've also got an intranet project where there is possible confusion to - 'am I searching the intranet, or am I searching all of EMBL?'

During the VF catch up with Cindy and Niki I noticed that govuk makes use of the placeholder text for search to define 'where' you are searching.

![Screenshot 2021-02-24 at 19 11 24](https://user-images.githubusercontent.com/925197/109053182-579d4100-76d4-11eb-975f-82577b926064.png)
![Screenshot 2021-02-24 at 19 11 41](https://user-images.githubusercontent.com/925197/109053167-52d88d00-76d4-11eb-93ea-f28cd0df6b20.png)

I think for search things, whether it's interactive or not - we should make use of the placeholder text as 'search is special'.

The next question to ask is, what would be the function of the interactive search filter when JavaScript breaks. Going back to govuk - they hide the search and show the sitemap:

![Screenshot 2021-02-24 at 19 17 29](https://user-images.githubusercontent.com/925197/109053979-3b4dd400-76d5-11eb-835f-ed7db37c9868.png)



In terms of the 'design' - I think this one matches better the stroke we have with existing icons we are using in `2.x`

![image](https://user-images.githubusercontent.com/925197/109054146-751eda80-76d5-11eb-8d18-a03cf607f33a.png)

**Comment by gabsie** [comment]
Thanks for this, @sturobson @nikiforosk @cindyebi

Yes, if I get this right, we are doing this in our case.. We will be saying Search in Projects
Looks good.

Will it go somewhere as a component, so that I refer the dev to it? :)
Gabs

**Reply by sturobson** [reply]
> Will it go somewhere as a component, so that I refer the dev to it? :)

Yes, @cindyebi is working on this in Figma at the moment too.

Hopefully we will get it into the code early next week I think,

**Comment by cindyebi** [comment]
I made the mockups for the default and filled variants for Search @sturobson :

![Screenshot 2021-02-26 at 14 15 51](https://user-images.githubusercontent.com/75834903/109311951-5be56d80-783e-11eb-8f53-ee8ab440316e.png)

**Reply by gabsie** [reply]
hey @cindyebi and @sturobson Stu - just wondered. You know how sometimes  (not in my current scenario, but yes to another scenario I have in mind), these also show suggestions below this search box? Is that something we want to show as well?

**Reply by sturobson** [reply]
Hey @gabsie - yes, I think it's something we need to define in the figma kit / code so that it's consistent across projects using it.

**Comment by cindyebi** [comment]
Hi all, there are two types of Search variants so far (see mockup below):
1. Default Search: a field form + a blue search button
2. Filtered Search: a field form without the blue search button; search suggestions will appear when a user is typing on the search field. 

![Search](https://user-images.githubusercontent.com/75834903/111477182-c34c5a00-8726-11eb-814b-4dade3219f47.png)

**Reply by nikiforosk** [reply]
I think that it will be helpful to explain in a bit more detail the context(s) in which each search box could be used.

1. Version 1 (with the button) assumes that there is enough space for the button. You can submit a keyword by pressing the button or by hitting enter. There could be suggestions (either as autocomplete or autosuggest [1]).

An example of that is the search box [on the training site](https://www.ebi.ac.uk/training/), especially when it is viewed on a desktop device:

![Screenshot 2021-03-18 at 09 14 12](https://user-images.githubusercontent.com/20280139/111601521-57233200-87ca-11eb-80dc-bacdacfcd31d.png)

 2. Version 2 (without the button) is probably meant to be used when there isn't enough space for the button or if you want to differentiate this search from another (more generic?) search box. The keyword is submitted by pressing enter (given that when the magnifier icon appears on the left of the box, it is typically used as a signifier instead of a button). There could be suggestions in this case too (either as autocomplete or autosuggest [1]).

I think we discussed a few contexts for that, including filtering a table. For example, [in Open Targets they are using a similar search box](https://beta.targetvalidation.org/target/ENSG00000142192/associations) as an instant filter of diseases that can be associated with a gene:

![Screenshot 2021-03-18 at 09 21 22](https://user-images.githubusercontent.com/20280139/111602722-8dad7c80-87cb-11eb-8494-f18fc43d767c.png)

In that case, the table is filtered as you type (i.e. you don't need to press enter to submit the query) so they are not showing suggestions.

There is another search box on the masthead of this page for the main search of the site (where you can search for genes and find associated diseases and vice versa) which looks and behaves differently. 

![Screenshot 2021-03-18 at 09 27 16](https://user-images.githubusercontent.com/20280139/111604062-d74a9700-87cc-11eb-828b-9f958728244e.png)

Another example is [searching within the online tutorial](https://www.ebi.ac.uk/training/online/courses/bioinformatics-terrified/search/page?s=bioinformatics), especially if we replace the link to "Search within this course" on the navigation panel on the left with a search box. In that case we'll need to use version 2 because there isn't enough space for a button in the navigation panel:

![Screenshot 2021-03-18 at 09 33 53](https://user-images.githubusercontent.com/20280139/111604453-41633c00-87cd-11eb-8af2-bd47d66ef730.png)

A third example would be viewing a site on mobile. The current search box both on the training home page and the search results within the online tutorial looks a bit disproportional:

![Screenshot 2021-03-18 at 09 36 46](https://user-images.githubusercontent.com/20280139/111604844-a0c14c00-87cd-11eb-8318-ad9a0522850c.png)

![Screenshot 2021-03-18 at 09 36 31](https://user-images.githubusercontent.com/20280139/111604846-a0c14c00-87cd-11eb-8a38-9eac5b947ae4.png)
 
A version with a magnifier glass may be more appropriate, although if we then have similar search boxes for different types of searches on the same page (as it happens with the Open Targets page) we may need to use some guiding text to explain what is searched in each case as @sturobson suggested or rely on where the box is placed: e.g. placing the filtering search box within the Table tab in Open Targets seems to signify that it searches the table underneath it whereas the search box on the masthead usually implies a search across the whole site.

[1] As I mentioned above, both boxes could offer suggestions or support instant search if this is appropriate in the context in which they are used. There are different kinds of suggestions, as explained in [this article by Tony Russell-Rose](https://uxmag.com/articles/designing-search-as-you-type-suggestions) (which also includes examples of instant search):

"Both auto-complete and auto-suggest offer varying degrees of support for query creation and reformulation, and the terms are used somewhat interchangeably by many people. But if we were to draw a precise distinction, it could be this:

- The purpose of auto-complete is to resolve a partial query, i.e., to search within a controlled vocabulary for items matching a given character string.

- The purpose of auto-suggest is to search a virtually unbounded list for related keywords and phrases, which may or may not match the precise query string."

I hope that this is helpful.

**Reply by sturobson** [reply]
the third example is fixed with https://latest.visual-framework.dev/components/vf-search/#vf-search--inline (to be deployed to npm)

**Comment by gabsie** [comment]
hey @cindyebi  - these look great.
I am only wondering a little bit about the contrast of the white on blue background, but maybe you've checked this.
looking forward to using these  - I need them for one of my designs :)
gabs

**Reply by nikiforosk** [reply]
In case this is of any help, in the training site we use a light gray background to highlight the suggestion that has been selected:

![Screenshot 2021-03-18 at 08 58 33](https://user-images.githubusercontent.com/20280139/111600110-f6472a00-87c8-11eb-9afa-428c3c6a0a3f.png)

Google seems to use a similar background (although the exact colour and the type they use are slightly different):

![Screenshot 2021-03-18 at 09 01 03](https://user-images.githubusercontent.com/20280139/111600387-4625f100-87c9-11eb-8b11-a4dc14b9a385.png)

**Reply by sturobson** [reply]
The colour contrast for white text on the interactive blue passes at 5.08

**Comment by cindyebi** [comment]
@gabsie Good catch Gabby, I did check the contrast yesterday and it passed all the other standards except for "Small texts AAA". 

@nikiforosk Since our interactive color is blue, I would suggest using a light blue option (included in our new color palette) for the selection:

![Screenshot 2021-03-18 at 09 57 43](https://user-images.githubusercontent.com/75834903/111607686-64dbb600-87d0-11eb-88fe-d3a2792f6fe9.png)

We could either:
1. Enlarge the text which is currently 16px so that it is accessible, keeping the interactive blue (Blue 600) color as a selection color.
2. Keep the text size as it is (16px) and change the selector color to light blue as it is our interactive color.

An example of how the gov.uk does it:
![Screenshot 2021-03-16 at 11 04 10](https://user-images.githubusercontent.com/75834903/111607952-a3717080-87d0-11eb-9013-4c86f774514b.png)

@nikiforosk The search field within the homepage of the training site will need an update since the label of the search field is no longer mandatory to be right on the top of the field as it is now. I had a separate trello card on the TOL two sprints ago on separating the Search label and the field. The rationale is the label isn't the same as the form field labels as Search is 'special'.

How the search label behaves on embl.org currently:
![Screenshot 2021-03-18 at 10 03 21](https://user-images.githubusercontent.com/75834903/111608491-2db9d480-87d1-11eb-82d7-4df2263ceb05.png)

I think we could add the 'How to Use' instructions on the VF site since it's more accessible to have these instructions on the site homepage than on Figma. 

It might be redundant to have the blue search button next to a Filtered Search because when a user has found what they are looking for from the suggestion, they might not need to click on the button anymore.

As for the mobile version, I think @sturobson already worked on the version of where the button and the field form are stacked. It is common practice that all buttons will stack and expand to fill in the screen on mobile, some examples:
Ebay desktop version:
![Screenshot 2021-03-18 at 10 08 58](https://user-images.githubusercontent.com/75834903/111609198-f7308980-87d1-11eb-8b75-da6b48fcc3a6.png)

Ebay mobile version:
![REFURBISHED-Vax-OnePWR-Blade-4-Cordless-Vacuum-Cleaner-0-6L-18V-CLSV-B4KSRB-5012512171883-eBay](https://user-images.githubusercontent.com/75834903/111609245-01eb1e80-87d2-11eb-9ea0-92624873ad28.png)

FT desktop version:
![Subscribe-to-read-Financial-Times](https://user-images.githubusercontent.com/75834903/111610155-ef251980-87d2-11eb-9698-4ce0c8be8582.png)

FT mobile version:
![Subscribe-to-read-Financial-Times (1)](https://user-images.githubusercontent.com/75834903/111610193-f9dfae80-87d2-11eb-831e-2ffdedd1a466.png)

This is because, on mobile, the bigger the 'target area' the better https://www.nngroup.com/articles/touch-target-size/

**Comment by cindyebi** [comment]
Stacked Search button from @sturobson @nikiforosk :

![Screenshot 2021-03-05 at 13 11 22](https://user-images.githubusercontent.com/75834903/111611622-922a6300-87d4-11eb-9811-aaa6422fcc73.png)

**Reply by sturobson** [reply]
This happens automatically -- if the search input has less than 60% (an arbitrary value I picked) of the width of it's container (in this case the sidebar column) it stacks automagically.

**Comment by sturobson** [comment]
After patch-y discussions yesterday I think I'm leaning towards:

- the autocomplete/autosuggest pattern being available where needed.
- always having a search button (which removes the magnifying glass'd 'filter' variant.

Thinking on this more:

- I don't think we can expect all site visitors to know that they can hit enter to submit the search.
- That JS will also work or always work correctly enough so that the autosuggest works - so a fallback would be to have the button.
- Original use cases for only having an input (where space was minimal) have mostly been mitigated (not having a search option in the sidebar on training - but a link to the search itself).
- The search will default to having the button drop under the input when there is not enough space.

**Reply by nikiforosk** [reply]
Hi Stu, I think that there is still an need for a search box without a big search button, e.g. when you need to filter a table, filter search results and also when there isn't enough space for the search button to appear next to the input field (e.g. in a mobile context). 

@gabsie and Carla have examples of tables which need to be filtered. As far as I understand these are similar to [the instant search on the table in Open Targets](https://beta.targetvalidation.org/target/ENSG00000171862/associations) that I shared yesterday ([under point no 2](https://github.com/visual-framework/vf-core/discussions/1388#discussioncomment-497412)).

Also in the training site, we may want to use this smaller search box to filter [the listings of events](https://www.ebi.ac.uk/training/live-events) or [on-demand training](https://www.ebi.ac.uk/training/on-demand): At the moment we're pointing to the main search as a workaround but user feedback has indicated that people are expecting to be able to filter within the listings page similarly to [how you're filtering events in embl.org](https://wwwdev.embl.org/events).

We suggested the link to the search page as an interim solution to searching within the online tutorial but we said that we'll revisit this once the smaller version of the search box is available.

Additionally, in a mobile context if you move the large button under the search box, the suggestions appear above it as it happens on the training site:

![Screenshot 2021-03-19 at 08 52 01](https://user-images.githubusercontent.com/20280139/111756933-adf43e80-8892-11eb-8e24-6baffc5c624a.png)

The search button cannot be used in that case, you can only select a suggestion or hit enter to submit your query.

This means that we'll need the variant with the magnifying glass if we prefer to have a button in that context. I have the impression that the magnifying glass originated from these contexts where there wasn't enough space for a search button.

I think that we do need a smaller search box (or more specifically a search box that doesn't need to be accompanied by a big search button) for these contexts. It may have the magnifying glass on the right as a button (in which case it will behave like the default version with the larger button) or on the left as a signifier (in which case you can submit your query with Enter). [As Tony Russell-Rose mentions in his article](https://uxmag.com/articles/designing-search-entering-the-query) the magnifying icon as button and submitting a query with Enter are well-known conventions for at least 10 years.

I our how to guide, we can explain that the version with the search button should be the default variant when searching is one of the main things that we expect people to do on the site and there is enough space on the page to accommodate the button (we can point to the training home page as an example of this).

The smaller variant with the magnifying glass is for special cases such as searching within a table, filtering search results, searching within sub-content (e.g. searching within the online tutorial) and when there isn't enough space for the large search button to appear next to the input box.

I hope this makes sense.

**Reply by sturobson** [reply]
when I worked on the search functionality for UCL - the company working on the back-end (Elastisearch) recommended that autosuggest on mobile should be hidden -- with the keyboard on screen you've not really got much space for autocomplete

**Reply by sturobson** [reply]
The other thing, which you don't see unless you're on an actual device, is that the OS will give you the correct keyboard, if the form is labeled correctly -- which I'd suggest voids the issue of the overlay here.

Obviously you don't get a special 'search' button on desktop OSs.

iOS:

![IMG_2229EA70C87F-1](https://user-images.githubusercontent.com/925197/111763551-31655e00-889a-11eb-803b-8cad2a3430c1.jpeg)

Android:

![Screenshot 2021-03-19 at 10 03 23](https://user-images.githubusercontent.com/925197/111763750-64a7ed00-889a-11eb-8b14-4f77f839161a.png)

**Reply by nikiforosk** [reply]
In a mobile context, e.g. on the BBC sports site, the search box tends to appear on the top of the screen to leave some space for the suggestions:

![CE78D000-F55B-45FF-A603-F232274EB4A2_1_105_c](https://user-images.githubusercontent.com/20280139/111763697-55c13a80-889a-11eb-88fe-470d40777cc3.jpeg)

Perhaps this can happen when the input field is in focus? 

Also the enter key on the keyboard is often labeled as Search. 

I think hiding the suggestions in order to be able to show the large search button under the input key will affect negatively the user experience, particularly in the training site. The suggestions are key to guiding learners on what content is available and help them search more efficiently, especially on a small screen.

**Reply by nikiforosk** [reply]
This is how the search box on the training looks like on my iphone:

![2AD0D384-66BE-4B1D-90FE-572117C0CE07](https://user-images.githubusercontent.com/20280139/111764630-4e4e6100-889b-11eb-9616-829dc026f88d.png)

As you mentioned, the search button under the input field is not usable but I can select a suggestion (which will submit it as a query) or hit return (which will select the first suggestion in this example). Notably, the button next to space on the keyboard isn't labelled search in this case (unlike the BBC site above).

**Reply by sturobson** [reply]
> I think hiding the suggestions in order to be able to show the large search button under the input key will affect negatively the user experience, particularly in the training site.

As mobile/tablet OSs have thought about this -- I think you can have the autocorrect -- but you would need to limit the results too I think - which is a suggestion I read on a UX site about this -- but cannot find.

**Reply by sturobson** [reply]
> Perhaps this can happen when the input field is in focus?

to be tested -- but I think the browser on mobile/tablet OSs would do this - if not … I don't think it's a good practice to move the site visitors position because if they tap out of the form - their position would be different?

**Comment by sturobson** [comment]
some points (in quote marks) from a [book of form design patterns](https://www.smashingmagazine.com/printed-books/form-design-patterns/):

As we are hiding the label for search (because it's special) the button becomes "a quasi-label for sighted users".

If we remove the button we "can't use it as a quasi label, the label would need to be reinstated, taking up room again" and "the button is a fundamental part of a form -- without it, it's not clear how users are meant to perform the search. While _we_ may be aware that forms can be submitted implicitly […] not all users are"

"Interestingly, many of the sites that **omit the submit button normally find room to include a magnifying glass icon** to signify its otherwise hidden affordance." (emphasis mine).

**Reply by nikiforosk** [reply]
I'm not really sure how the search field works when voice offer is on but I can ask my blind friends. I would somehow expect it to identify itself as "search"... Wouldn't the magnifying icon be read out as "search button" on voice over?

The version of the search box with the icon on the left as a signifier is really meant to support instant search like the filtering of the table in Open Targets where you don't really need to submit the search query. I don't know how accessible this is either. I agree with you that it's probably better to have a button there too, except that in my opinion the input field there should be smaller and the button should look as like magnifying glass so that it fits on the same row (when there isn't much space for a larger button).

**Reply by sturobson** [reply]
VO on MacOS latest, when tabbing through, will:

- read out the label (there is a hidden label)
- read out the placeholder
- read out the button

**Reply by sturobson** [reply]
for reference here's a (five year old) post on what some assistive technologies announce depending on how the search form is marked up

https://www.a11ymatters.com/pattern/accessible-search/

**Comment by cindyebi** [comment]
@nikiforosk @sturobson and I discussed the Search field variants on the vf catch-up meeting the other day and we decided to add the magnifying icon button as a replacement for the large "Search" button as a responsive mobile variant. Basically, the Search label would 'fade' out and be replaced with an icon once the screen is resized. Mockups below:
1. Responsive Large: for screens >600px width
<img width="826" alt="Screenshot 2021-03-26 at 16 04 59" src="https://user-images.githubusercontent.com/75834903/112660702-a104bb80-8e4d-11eb-9481-5735edc277ac.png">

2. Responsive Small: for screens <600px width
<img width="820" alt="Screenshot 2021-03-26 at 16 05 04" src="https://user-images.githubusercontent.com/75834903/112660712-a3671580-8e4d-11eb-9f47-988a55ddf22a.png">

@sturobson has a demo video of how it works exactly.

**Comment by khawkins98** [comment]
PR for this at https://github.com/visual-framework/vf-core/pull/1455/
